[18.0][IMP] account_payment_batch_oca: add mail notification

alexis-via opened the pull request:
Mail notification is a basic feature so it should be directly available in account_payment_batch_oca. I tried to make it as simple as simple possible, both for the user and in the code. Mail notification must be enabled on the payment method ; if enabled, mails are sent when state goes from "generated" to "uploaded" (final step).

The mail template gives the payment date, label, amount and bank account for each account.payment. If the payment is composed of several transactions, it gives the detail of the transactions (amount and communication), to make it easy for the partner who receives the email to have all the details of the payments.

![mail_notif](https://github.com/user-attachments/assets/9b822989-59cb-4956-96ef-0bc767ae6813)

alexis-via committed:
[IMP] account_payment_batch_oca: Add mail notification when payment/debit order arrives on 'uploaded' state

Update readme.

alexis-via committed:
[FIX] account_payment_sepa_direct_debit: fix generation of debit orders when payment line is created manually

alexis-via committed:
[IMP] enable mail notification in all tests

pedrobaeza commented:
This already exists in https://github.com/OCA/bank-payment/tree/16.0/account_payment_order_notification

alexis-via committed:
[IMP] Improve variable names and re-organise SEPA generation code

Several variable names haven't been changed since the adoption of account.payment. Having good variable names is important to understand the code.

alexis-via committed:
[IMP] account_payment_batch_oca: fix and improve test

Remove demo data in tests
Avoid error on supplier refund confirmation "odoo.exceptions.UserError: This move
is configured to be auto-posted on 03/07/2025"

alexis-via commented:
I know, I studied account_payment_order_notification and I came to the conclusion that it was possible to develop something more configurable with less code and no specific object (account.payment.order.notification). I'm happy with the result so far. I also took the opportunity to improve the content of the email. And I think it's better to provide this feature directly in account_payment_batch_oca because it should be considered as a core feature (and more users will see the feature and use it !).

BTW, working on the code these last days, I figured out that the big cleanup/rewrite of the SEPA XML generation code is not finished and there is still room for improvement. So I continue to work on that too.

alexis-via committed:
[IMP] Continue code cleanup in SEPA XML file generation

Reduce number of args in methods
Improve variable names to avoid mis-leading names
Remove @api.model where it's not appropriate
Add raise_if_oversized=True in _prepare_field() where appropriate

alexis-via committed:
[IMP] Re-structure the code to generate SEPA Creditor Identifier

Don't set SchmeNm/Prtry to "SEPA" when the payment order is not SEPA.

alexis-via committed:
[IMP] Mutualize code for PmtId XML block between SCT and SDD

alexis-via committed:
[IMP] Mutualize InstdAmt tag between SCT and SDD

alexis-via committed:
[IMP] Update and remove comments

alexis-via commented:
/ocabot merge major
